### PR TITLE
Fix ranges::find_end return for bidi-common case

### LINKED ISSUE #2268: <algorithm>: The `bidirectional common range` version of `ranges::find_end` returns the wrong value
https://github.com/microsoft/STL/blob/88332706f49d3297b200b5605ef174fc47d38d1a/stl/inc/algorithm#L2361-L2372

IIUC, the first return statement should return `{std::move(_Next1), std::move(_Candidate)};`.

Test case:
```
#include <ranges>
#include <algorithm>

constexpr auto find = [] {
  auto r = std::views::iota(0, 2)
         | std::views::filter([](int) { return true; });
  auto s = std::ranges::find_end(r, std::views::single(1));
  return s.front() == 1;
}();

static_assert(find);
```
https://godbolt.org/z/1dzq8Pe1K

### PR DESCRIPTION
**Pull request by CaseyCarter**
...and add full test coverage.

Fixes #2268

### CODE REVIEW

**Review by miscco — APPROVED** [review]
Obligatory "everything that is not testen is broken"

File: tests/std/tests/P0896R4_ranges_alg_find_end/test.cpp [hunk 1]
```diff
@@ -11,104 +11,98 @@
...
-constexpr void smoke_test() {
-    using ranges::find_end, ranges::iterator_t, ranges::subrange, std::array, std::same_as;
-    using P = std::pair<int, int>;
+using namespace std;
 
-    // Validate dangling story
-    STATIC_ASSERT(same_as<decltype(find_end(borrowed<false>{}, array<int, 42>{})), ranges::dangling>);
-    STATIC_ASSERT(same_as<decltype(find_end(borrowed<true>{}, array<int, 42>{})), subrange<int*>>);
+// Validate dangling story
+STATIC_ASSERT(same_as<decltype(ranges::find_end(borrowed<false>{}, array<int, 42>{})), ranges::dangling>);
```
Comment (miscco) [inline tests/std/tests/P0896R4_ranges_alg_find_end/test.cpp:17]: Nitpick, we could get ein

File: tests/std/tests/P0896R4_ranges_alg_find_end/test.cpp [hunk 2]
```diff
@@ -11,104 +11,98 @@
...
-        assert(result.begin() == pairs.begin() + 4);
-        assert(result.end() == pairs.begin() + 6);
-    }
+        {
+            // Validate iterator + sentinel overload [found case]
+            Fwd1 haystack{pairs};
+            Fwd2 needle{good_needle};
+            const same_as<subrange<iterator_t<Fwd1>>> auto result =
+                find_end(begin(haystack), end(haystack), begin(needle), end(needle), pred, get_first);
+            STATIC_ASSERT(
```
Comment (miscco) [inline tests/std/tests/P0896R4_ranges_alg_find_end/test.cpp:63]: Nitpick, is there a reason to shout `STATIC_ASSERT`, at least in ranges code we should be able to be a bit less shouty

File: tests/std/tests/P0896R4_ranges_alg_find_end/test.cpp [hunk 3]
```diff
@@ -11,104 +11,98 @@
...
+    const int haystack[]                                    = {1, 2, 3, 1, 2, 3, 1, 2, 3};
+    const int needle[]                                      = {2, 3, 1};
+    const same_as<ranges::subrange<const int*>> auto result = ranges::find_end(haystack, span<const int>{needle});
+    assert(result.begin() == haystack + 4);
+    assert(result.end() == haystack + 7);
+    return true;
 }
 
+int main() {
 #ifndef _PREFAST_ // TRANSITION, GH-1030
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply tests/std/tests/P0896R4_ranges_alg_find_end/test.cpp:102]: Running all 17 test configurations on my dev box takes 38 seconds. If I change this to `#if 1`, it takes 89 seconds. The two `/analyze` configs are _very_ long poles.

On the bright side, they do at least finish now. IIRC I've seen a couple of fixes for memory leaks in the static analyzer in the last year which seem to have corrected the memory consumption issues we were having. Despite being long-run, the analyzer configs _do_ complete.

File: tests/std/tests/P0896R4_ranges_alg_find_end/test.cpp (hunk 2 above)
Comment (CaseyCarter) [reply tests/std/tests/P0896R4_ranges_alg_find_end/test.cpp:63]: Yes there is: I refuse to manually map from line number to failure condition when I can make MSVC do it for me by capitalizing `static_assert`.

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed, or if more work is required.

**Review by mnatsuhara — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks for finding the end of this bug! :lady_beetle: :stop_sign: :joy_cat:

**Pull request merged**